### ISSUE TITLE
Spike how to query sensitive data

### ISSUE THREAD

**Issue by jon-betts** [hypothesis/report#118]
This is the first time we will have to present sensitive data which cannot be stored at rest on the DB we are querying directly.

Our plan is to:

 * Have a non-materialized view which unions the two tables
 * Query over that

However we need to know that this will be performant enough. We need a plan which:

 * Is sufficient for tables of the size of the user table (~250k) + room for growth
 * Ideally is sufficient for tables of any size

## Postgres can push down where clauses into views sometimes

Under certain circumstances Postgres is smart enough to push down where clauses into views: https://www.postgresql.org/message-id/3248.1194140315@sss.pgh.pa.us

We could do with finding out:

 * Will this work in our case (with FDW and imagined region columns)?
 * Does using a region column create trouble, or can we filter on ids without specifying region and then filter them out later?
 * Will Metabase ruin all of this if we make it a model?

In order to do this we will need to use explain on the real machine to see Postgres' query plan. You can't get any sensible timing from Metabase as the interface is too slow. We can use the `event` table as a proxy with many more rows. If it works for that, it's probably good.

Some sketchy details about getting onto the report box with `psql` here: https://hypothes-is.slack.com/archives/C042JM4SPD0/p1666796086327609

## Tasks

 * https://github.com/hypothesis/lms/pull/4766 
    * This has been replaced by: https://github.com/hypothesis/lms/issues/4730

**Comment by jon-betts** [comment]
I've tried this out using the `report.events` table from LMS as a proxy. It has roughly twice the the number of rows as the LMS users table, so is a reasonably proxy we already have access to.

To do this I constructed a fake users view which used:

 * `organization_id` - as a proxy for `id`
 * `region` - A static string (`ca` or `us` as appropriate) - Only exists on our side, not remote
 * `id` - `region` and `organization_id` concatenated

As `id` is just `region` and `organization_id` concatenated, the following SQL statements are equivalent:

 * `SELECT COUNT(1) FROM lms.users_test WHERE id='us-136';`
 * `SELECT COUNT(1) FROM lms.users_test WHERE organization_id=136 AND region='us';`
 
The results of analzying the execution are:

```sql
EXPLAIN ANALYZE VERBOSE SELECT COUNT(1) FROM lms.users_test WHERE id='us-136';

-------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=458.64..458.65 rows=1 width=8) (actual time=11507.738..11507.740 rows=1 loops=1)
   Output: count(1)
   ->  Append  (cost=100.00..458.56 rows=34 width=0) (actual time=2.656..11507.086 rows=3965 loops=1)
         ->  Foreign Scan on lms_us.events  (cost=100.00..229.19 rows=17 width=0) (actual time=2.654..7505.353 rows=3965 loops=1)
               Filter: (concat('us-', events.organization_id) = 'us-136'::text)
               Rows Removed by Filter: 610443
               Remote SQL: SELECT organization_id FROM report.events
         ->  Foreign Scan on lms_ca.events events_1  (cost=100.00..229.19 rows=17 width=0) (actual time=4001.344..4001.344 rows=0 loops=1)
               Filter: (concat('ca-', events_1.organization_id) = 'us-136'::text)
               Rows Removed by Filter: 4937
               Remote SQL: SELECT organization_id FROM report.events

 count 
-------
  3965
(1 row)

Time: 12284.475 ms (00:12.284)
```

Vs.

```sql
EXPLAIN ANALYZE VERBOSE SELECT COUNT(1) FROM lms.users_test WHERE organization_id=136 AND region='us';
                                                           QUERY PLAN                                                            
---------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=153.13..153.14 rows=1 width=8) (actual time=132.534..132.536 rows=1 loops=1)
   Output: count(1)
   ->  Append  (cost=100.00..153.09 rows=17 width=0) (actual time=3.268..132.221 rows=3965 loops=1)
         ->  Foreign Scan on lms_us.events  (cost=100.00..153.00 rows=17 width=0) (actual time=3.266..131.891 rows=3965 loops=1)
               Remote SQL: SELECT NULL FROM report.events WHERE ((organization_id = 136))
 Planning Time: 0.112 ms
 Execution Time: 134.864 ms

 count 
-------
  3965
(1 row)

Time: 116.843 ms
```

## Analysis

In the first case we can see that a full retrieval of the table is done, which we then filter down. This is kind of the worst case scenario: we retrieve every row and then throw almost all of them away. This is because Postgres has to construct the composite column we create on ourside, and isn't smart enough to work out how to unpick this to do it efficiently.

In the other instance using region and organization id directly we see a much better outcome. Two interesting things have happened here:

 * Postgres has realised that the static region value means half of the union does not apply. This is only issuing a query to the US LMS, not the CA one
 * The where clause for fields which are directly copied from the underlying DB are pushed down into that DB. The filtering on organization id is happening in LMS, not in report

## Conclusion

Both of these statements are logically equivalent and return the same information, but in one case it returns 1000x quicker.

In summary:

 * This looks like it will work
 * You must not filter or join on composite fields
 * Conditions on fields which are directly mapped from the underlying values can be pushed down the the remote DB
 * Postgres can intelligently prune `UNION` branches based on static values
